### PR TITLE
feat(backend): Introduce handshake & extract logic auth logic to separate classes

### PR DESCRIPTION
**Pull request by dimkl**
## Changes made in this PR

- Drop `Clerk::SDK#interstitial` method (it should not be used by any customer since this is internal)
- Replace the interstitial mechanism with the handshake mechanism
- Extract logic from the RackMiddlewareV2 to `AuthenticateRequest` class
- Extract data used from the request and the configuration to `AuthenticateContext` class ([Parameter object pattern](https://refactoring.guru/introduce-parameter-object))
- Introduce constants module to centralize and re-use the constants value used across the SDK

## Customer impact
- Interstitial HTML page rendered when we cannot determine the request state is replaced with 2 HTTP 307 redirects
- Introduced `CLERK_PUBLISHABLE_KEY` env which is required for the handshake to work properly
- `CLERK_SECRET_KEY` should use the format `sk_test | sk_live` format

## TODO
- Add tests
- Check all uses cases

## Next Steps

- Support multi-domain logic

### CODE REVIEW

File: lib/clerk/authenticate_context.rb [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    
+        def browser_request?
+            @headers.user_agent && @headers.user_agent.starts_with?("Mozilla/")
+        end
+
+        def dev_browser?
+            !dev_browser.empty?
+        end
+
+        def header_token?
```
Comment (SokratisVidros) [inline lib/clerk/authenticate_context.rb:167]: This method name is a bit misleading. It feels simpler to have two methods such as `session_token_in_header`, `session_token_in_cookie` and work with them.

File: lib/clerk/authenticate_request.rb [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            end
+        end
+
+        protected
+
+        def resolve_header_token(env)
+            begin
+                # malformed JWT
+                return signed_out(reason: TokenVerificationErrorReason::TOKEN_INVALID) if !sdk.decode_token(auth_context.header_token)
+            rescue JWT::DecodeError
```
Comment (SokratisVidros) [inline lib/clerk/authenticate_request.rb:37]: Please combine all the rescue in a single begin/end.

File: lib/clerk/authenticate_request.rb [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            end
+
+            # Clerk.js should refresh the token and retry
+            return signed_out(enforce_auth: true)
+        end
+
+        def resolve_cookie_token(env)
+            # in cross-origin XHRs the use of Authorization header is mandatory.
+            # TODO: add reason
+            return signed_out if auth_context.cross_origin_request?
```
Comment (SokratisVidros) [inline lib/clerk/authenticate_request.rb:58]: ❓ Shouldn't this check go to the resolve_header_token method?

File: lib/clerk/authenticate_context.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                handshake_token: request.cookies[HANDSHAKE_COOKIE],
+                dev_browser: request.cookies[DEV_BROWSER_KEY]
+            })
+
+            header_token = request.env[AUTHORIZATION_HEADER] || ''
+            @headers = OpenStruct.new({
+                header_token: header_token.strip.sub(/\ABearer /, ''),
+                sec_fetch_dest: request.env[SEC_FETCH_DEST_HEADER],
+                accept: request.env[ACCEPT_HEADER] || "",
+                user_agent: request.env[USER_AGENT_HEADER] || "",
```
[the remark this answers is not in the document]
Comment (dimkl) [reply lib/clerk/authenticate_context.rb:41]: no we don't. i will drop it.

File: lib/clerk/authenticate_request.rb (hunk 3 above)
Comment (dimkl) [reply lib/clerk/authenticate_request.rb:58]: Since the condition used to execute the `resolve_header_token` flow is the existence of a `header_token` and the case described here is that we want to return a `signed_out` state if it's a cross-origin request and there is no header_token, it will never be triggered in the `resolve_header_token` flow.

File: lib/clerk/authenticate_context.rb [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+        def instance_type
+            @instance_type ||= publishable_key.start_with?("pk_live_") ? "production" : "development"
+        end
+
+        def development_instance?
+            # TODO(dimkl): Consolidate with @clerk/backend to use pk instead of sk
+            # instance_type == "development"
+
+            secret_key.start_with?("test_") || secret_key.start_with?("sk_test_")
```
[the remark this answers is not in the document]
Comment (dimkl) [reply lib/clerk/authenticate_context.rb:115]: dropped the pre-kima format

File: lib/clerk/authenticate_context.rb [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+            elsif instance_type == "development" && domain
+                "clerk.#{domain}"
+            else
+                # remove $ postfix
+                decode_publishable_key(publishable_key).chop
+            end
+        end
+
+        def instance_type
+            @instance_type ||= publishable_key.start_with?("pk_live_") ? "production" : "development"
```
[the remark this answers is not in the document]
Comment (dimkl) [reply lib/clerk/authenticate_context.rb]: i removed `instance_type`. I will depend on the `secret_key` values to identify the instance type.

**Review by SokratisVidros — APPROVED** [review]

File: lib/clerk.rb [hunk 7]
```diff
@@ -50,6 +51,8 @@ def initialize
         @api_key = secret_key
       end
 
+      @publishable_key = ENV["CLERK_PUBLISHABLE_KEY"]
```
Comment (adentaofftake) [inline lib/clerk.rb:54]: are these cannonical ENV's documented anywhere?

**Review by adentaofftake — COMMENTED** [review]
lot of this goes over my head. If it works, I'm excited!

File: lib/clerk.rb (hunk 7 above)
Comment (adentaofftake) [reply lib/clerk.rb:54]: yeah looks like this is new, might make sense to update the README 🤷‍♂️

File: lib/clerk/authenticate_context.rb [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+                host: request.host,
+                port: request.port
+            })
+        end
+
+        ##
+        # The following properties are part of the props supported in all the AuthenticateContext
+        # objects across all of our SDKs (eg JS, Go)
+        def secret_key
+            @config.api_key || ""
```
Comment (gkats) [inline lib/clerk/authenticate_context.rb:51]: :wink: Same for the rest.

```suggestion
            @config.api_key.to_s
```

File: lib/clerk/authenticate_context.rb [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        end
+
+        def clerk_redirect_url
+             # TODO(dimkl): Add multi-domain support
+             ""
+        end
+
+        ##
+        # Helper methods supported only by Ruby SDK to reduce code duplication and improve readability
+        #
```
Comment (gkats) [inline lib/clerk/authenticate_context.rb:87]: :question: Is this comment actually needed?

File: lib/clerk/authenticate_context.rb [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+        def dev_browser
+            @dev_browser ||= retrieve_from_query_string(@clerk_url, DEV_BROWSER_COOKIE) || @cookies.dev_browser || ""
+        end
+
+        # The frontend_api returned is without protocol prefix
+        def frontend_api
+            return "" if !publishable_key || !valid_publishable_key?(publishable_key)
```
Comment (gkats) [inline lib/clerk/authenticate_context.rb:95]: :upside_down_face: Perhaps the `valid_publishable_key?` method can also check that the key is not blank, eliminating the need to add this OR clause here.

Or we can just do 
```ruby
return "" if !valid_publishable_key?(publishable_key.to_s)
```

File: lib/clerk/authenticate_context.rb [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+      
+            origin != request_host
+        end
+
+        def dev_browser?
+            !dev_browser.empty?
+        end
+
+        def session_token_in_header?
+            session_token_in_header && !session_token_in_header.empty?
```
Comment (gkats) [inline lib/clerk/authenticate_context.rb:155]: Same for the rest.
```suggestion
            !session_token_in_header.to_s.empty?
```

File: lib/clerk/authenticate_request.rb [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+                raise e
+            end
+        end
+
+        def handle_handshake_maybe_status(env, **opts)
+            return signed_out if !eligible_for_handshake?
+            handshake(env, **opts)
+        end
+
+        # A outcome
```
Comment (gkats) [inline lib/clerk/authenticate_request.rb:169]: :question: What is this comment about? Perhaps we can abstract the possible outcomes in classes and make sure we return/invoke the correct outcome?

Would that convey the notion of an outcome better in code?

File: lib/clerk/authenticate_request.rb [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+            rescue JWT::ExpiredSignature
+                handshake(env, reason: TokenVerificationErrorReason::TOKEN_EXPIRED)
+            rescue JWT::InvalidIatError
+                handshake(env, reason: TokenVerificationErrorReason::TOKEN_NOT_ACTIVE_YET)
+            end
+
+            signed_out
+        end
+
+        def resolve_handshake(env)
```
Comment (gkats) [inline lib/clerk/authenticate_request.rb:114]: :cry: This method is a bit hard to follow. Might be just me.

File: lib/clerk/authenticate_request.rb [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+                sdk.verify_token(token, **opts)
+            rescue JWT::ExpiredSignature, JWT::InvalidIatError => e
+                raise e
+            rescue JWT::DecodeError, JWT::RequiredDependencyError => e
+                false
+            end
+        end
+
+        def sdk
+            Clerk::SDK.new
```
Comment (gkats) [inline lib/clerk/authenticate_request.rb:234]: :question: Should we memoize this?

**Review by gkats — APPROVED** [review]
Looks good to me overall, but I admit that the handshake implementation is a bit hard to follow. 

Might be just me though.

Great work on a tough problem!

File: lib/clerk/authenticate_request.rb (hunk 12 above)
Comment (gkats) [reply lib/clerk/authenticate_request.rb:169]: I saw after I finished the review that this code has been moved over, along with the comments.

Not suggesting you do anything at this point, unless you're up for a refactoring trip. :smile: :champagne:

File: lib/clerk/authenticate_context.rb (hunk 9 above)
Comment (dimkl) [reply lib/clerk/authenticate_context.rb:87]: i will drop it :)

File: lib/clerk/authenticate_request.rb (hunk 14 above)
Comment (dimkl) [reply lib/clerk/authenticate_request.rb:234]: I thought of doing that in a separate PR. I kept it the same with our current codebase.
@agis Could you remember if there is a reason behind not memorizing this?
Also since we are using a cache for the jwks keys which is shared across the SDK instances it seems okay to leave it as is.
```ruby
# lib/clerk/sdk.rb
@@jwks_cache = JWKSCache.new(JWKS_CACHE_LIFETIME)
```

File: lib/clerk/authenticate_request.rb (hunk 13 above)
Comment (dimkl) [reply lib/clerk/authenticate_request.rb:114]: Would comments help? The main idea here is that if you find the `__handshake_token` query parameter, then verify/decode the value (It's a JWT) and assign the cookie directives find in the decoded value as headers in the response. If there is no `__session` directive then it's signed-out request, else you verify the session token inside the `__session` directive. For development instances if the session token inside the handshake token is expired or not active yet, we try to verify it with a big leeway (1 day) to skip clock skew issues. If all of those fail, we raise an error because something does not work as expected in the handshake mechanism.

I don't know if the original implementation is easier to follow but this is the original implementation in JS i ported to Ruby: https://github.com/clerk/javascript/blob/8e5c881c40d7306c5dbd2e1f1803fbf75127bd71/packages/backend/src/tokens/request.ts#L92

File: lib/clerk/authenticate_request.rb (hunk 14 above)
Comment (agis) [reply lib/clerk/authenticate_request.rb:234]: Yep, this is deliberately not memoized, since we don't want it to be shared across requests. See https://github.com/clerk/clerk-sdk-ruby/commit/11239115d79603f5178bfdada9fe94369427818f.

File: lib/clerk/authenticate_request.rb (hunk 13 above)
Comment (gkats) [reply lib/clerk/authenticate_request.rb:114]: Thanks for the explanation! Comments do help, and I also think that extracting methods, renaming fields etc. will help.

This is definitely something that can be done at a later stage, don't let it delay your work.

File: lib/clerk/authenticate_request.rb (hunk 12 above)
Comment (dimkl) [reply lib/clerk/authenticate_request.rb:169]: i will keep it as is for now and we should tackle it in the future if there is a need to refactor it :)

File: lib/clerk.rb (hunk 7 above)
Comment (dimkl) [reply lib/clerk.rb:54]: i have updated the readme file to include this ENV in the configuration example. I have also added a section about the new mechanism we have introduced.

File: lib/clerk/authenticate_request.rb (hunk 12 above)
Comment (gkats) [reply lib/clerk/authenticate_request.rb:169]: Yessir!

File: lib/clerk/authenticate_request.rb (hunk 13 above)
Comment (dimkl) [reply lib/clerk/authenticate_request.rb:114]: i will add some comments and try to split this method to smaller building blocks to improve readability.

File: lib/clerk/authenticate_context.rb [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+        end
+
+        def development_instance?
+            # TODO(dimkl): Consolidate with @clerk/backend to use pk instead of sk
+            secret_key.start_with?("sk_test_")
+        end
+
+        def production_instance?
+            # TODO(dimkl): Consolidate with @clerk/backend to use pk instead of sk
+            secret_key.start_with?("sk_live_")
```
Comment (agis) [inline lib/clerk/authenticate_context.rb:111]: These may also start with `live_`

File: lib/clerk/authenticate_context.rb [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+        def valid_publishable_key_prefix?(pk)
+            pk.start_with?("pk_live_") || pk.start_with?("pk_test_")
+        end
+
+        def valid_publishable_key_postfix?(pk)
+            decode_publishable_key(pk).end_with?("$")
+        end
+
+        def decode_publishable_key(pk)
+            Base64.decode64(pk.split("_")[2].to_s)
```
Comment (agis) [inline lib/clerk/authenticate_context.rb:178]: 🔧 Shouldn't we raise an error if `pk.split("_")[2] == nil`?

File: lib/clerk/authenticate_context.rb [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        end
+
+        def retrieve_from_query_string(url, key)
+            request_qs = Rack::Utils.parse_query(url.query)
+            request_qs[key]
```
Comment (agis) [inline lib/clerk/authenticate_context.rb:183]: 🙃 would probably put this in a single line, without the intermediate variable

File: lib/clerk/authenticate_context.rb (hunk 15 above)
Comment (dimkl) [reply lib/clerk/authenticate_context.rb:111]: After discussing it with @SokratisVidros we considered dropping the legacy AuthV1 format. Also since we are gonna do a major bump for this change, i think we are okay. Does it make any sense? Do you have any considerations?

File: lib/clerk/authenticate_context.rb (hunk 16 above)
Comment (dimkl) [reply lib/clerk/authenticate_context.rb:178]: i forgot to validate the existence of the publishable key. I will add it in a separate PR with some comments and minor changes.

File: lib/clerk/authenticate_context.rb (hunk 17 above)
Comment (dimkl) [reply lib/clerk/authenticate_context.rb:183]: i will refactor it in the PR with the minor changes / fixes.
